File: k-closest-points-to-origin/k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<int[]>(Solution::customEucledianComparator);
        for (int[] point : points) {
            pq.offer(point);
        }
        List<int[]> result = new ArrayList<>();
        while (!pq.isEmpty() && k-- > 0) {
            result.add(pq.poll());
        }
        int[][] res = new int[result.size()][];
        for (int i = 0; i < result.size(); i++) {
            res[i] = result.get(i);
        }
        return res;
    }
    
    private static int customEucledianComparator(int[] a, int[] b) {
        double aToCenter = Math.sqrt(a[0]*a[0] + a[1]*a[1]);
        double bToCenter = Math.sqrt(b[0]*b[0] + b[1]*b[1]);
        return Double.compare(aToCenter, bToCenter);
    }
}

File: k-closest-points-to-origin/k-closest-points-to-origin.go
type CartesianPoints [][]int

func kClosest(points [][]int, k int) (result [][]int) {
    sort.Sort(CartesianPoints(points))
    i := 0
    for k > 0 && i < len(points) {
        result = append(result, points[i])
        k--
        i++
    }
    return
}

func (c CartesianPoints) Less(a, b int) bool {
    return gapToCenter(c[a]) < gapToCenter(c[b])
}

func (c CartesianPoints) Swap(a, b int) {
    c[a], c[b] = c[b], c[a]
}

func (c CartesianPoints) Len() int {
    return len(c)
}

func gapToCenter(a []int) float64 {
    return math.Sqrt(float64(a[0]*a[0] + a[1]*a[1]))
}